src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,634 @@
+    for (int i = static_cast<int>(indices.size()) - 1; i >= 0; --i) {
+        if (indices.at(i) != border - indices.size() + static_cast<std::size_t>(i)) {
+            index = static_cast<std::size_t>(i);
+            break;
+        }
+    }
+    indices.at(index)++;
+    for (std::size_t i = index + 1; i < indices.size(); ++i) {
+        indices.at(i) = indices.at(index) + i - index;
+    }
+}
+
+template <typename T>
+std::vector<std::vector<T>> GetSubsets(std::vector<T> const& elements, std::size_t const n) {
+    if (elements.size() < n) {
+        return {};
+    }
+    std::vector<std::vector<T>> result = {};
+    std::vector<std::size_t> indices = std::vector<std::size_t>(n);
+    std::iota(std::begin(indices), std::end(indices), 0);
+    while (!indices.empty()) {
+        std::vector<T> element = std::vector<T>(indices.size());
+        auto get_ith = [&elements](std::size_t i) { return elements.at(i); };
+        std::transform(indices.begin(), indices.end(), element.begin(), get_ith);
+        result.push_back(std::move(element));
+        NextSubset(indices, elements.size());
+    }
+    return result;
+}
+
+bool CheckToken(Token const& token, std::string& tok, graph_t const& graph, graph_t const& pattern,
+                Embedding const& embedding) {
+    if (token.first == -1) {
+        tok = token.second;
+    } else {
+        vertex_t u = boost::vertex(token.first, pattern);
+        vertex_t v = embedding.at(u);
+        std::map<std::string, std::string> attrs = graph[v].attributes;
+        if (attrs.find(token.second) == attrs.end()) {
+            return false;
+        }
+        tok = attrs.at(token.second);
+    }
+    return true;
+}
+
+bool Satisfied(std::vector<Literal> const& literals, graph_t const& graph, graph_t const& pattern,
+               Embedding const& embedding) {
+    for (Literal const& l : literals) {
+        Token fst_token = l.first;
+        Token snd_token = l.second;
+        std::string fst, snd;
+        bool first_res = CheckToken(fst_token, fst, graph, pattern, embedding);
+        bool second_res = CheckToken(snd_token, snd, graph, pattern, embedding);
+        if (!first_res or !second_res or fst != snd) return false;
+    }
+    return true;
+}
+
+bool Validate(graph_t const& graph, Gfd const& gfd, Embeddings const& embeddings,
+              std::size_t const sigma) {
+    bool result = false;
+    std::map<vertex_t, std::set<vertex_t>> maps = {};
+    for (Embedding const& embedding : embeddings) {
+        if (!Satisfied(gfd.GetPremises(), graph, gfd.GetPattern(), embedding)) {
+            continue;
+        }
+        result = true;
+
+        if (!Satisfied(gfd.GetConclusion(), graph, gfd.GetPattern(), embedding)) {
+            return false;
+        }
+
+        BGL_FORALL_VERTICES_T(v, gfd.GetPattern(), graph_t) {
+            std::set<vertex_t> vertices = {embedding.at(v)};
+            auto [it, emplaced] = maps.try_emplace(v, std::move(vertices));
```

Comment:
maybe i'm doing something wrong, but my code doesn't compile if i do this

Reply:
Yea, I apologize, it won't work directly like this. Still the suggested change makes sense since there's no point to create an object and move it where we can directly use it in the `emplace`.

The issue here is that `embedding.at(v)` is just a `vertex_t`, an integer. 
`auto [it, emplaced] = maps1.try_emplace(v, value);` -- `try_emplace` expects the second argument to be of type `std::set<vertex_t>`, but you're providing a `vertex_t` instead, and since `std::set<vertex_t>` doesn't have a constructor that takes a single `vertex_t` directly, it results in a compilation error.

Note that this is different from this: `std::set<vertex_t> vertices = { value };`, because that is a constructor from a `initilizer_list`.

So, in order to resolve that issue, you just need to use correct constructor, that is, `auto [it, emplaced] = maps1.try_emplace(v, std::set<vertex_t>{embedding.at(v)});`

https://godbolt.org/z/7zeGPcnx6